### PR TITLE
DynamicDependencies: Use Win11 APIs if available

### PR DESCRIPTION
**Pull request by DrusTheAxe**
DynamicDependencies: Use Win11 APIs if available

Leverage the the Win11 Dynamic Dependency API if available, otherwise use WinAppSDK's implementation (aka polyfill). The WinAppSDK DynDep C API handles this, thus benefiting all callers of WinAppSDK's C and WinRT DynDep API (e.g. C# via projections).

This happens on Win11 >= 22H2 (aka SV2).

NOTE: This makes the DDLM irrelevant on Win11 22H2+ (>SV2). A future (separate) PR will change the intaller to not install the DDLM packages on these systems.

*NOTE:* Persisted definitions aren't imported into the OS after OSupgrade to Win11 22H2+. If you create a persisted definition (LifetimeKind=FilePath|RegistryKey) on older you'll need to re-create them after the OS upgrades.

*NOTE:* MddGetResolvedPackageFullNameForPackageDependency()'s behavior differed from GetResolvedPackageFullNameForPackageDependency(). Both return null vs package full name when the PackageDependencyId is not resolved vs resolved but the HRESULT return value differed

|PackageDependencyId|WinAppSDK                         |OS API|
|-------------------|----------------------------------|------|
|unknown            |HRESULT_FROM_WIN32(ERROR_NO_MATCH)|S_OK  |
|not resolved       |S_OK                              |S_OK  |
|resolved           |S_OK                              |S_OK  |

MddGetResolvedPackageFullNameForPackageDependency() has been changed to match the OS API. A new OS API will be needed to distinguish between a known valid PackageDependencyId that's not been resolved yet vs an invalid/unknown PackageDependencyId.

NOTE: This change is compatible with the documented behavior. It'll be noted in the release notes.

NOTE: As a byproduct of return MddGetResolvedPackageFullNameForPackageDependency() error handling the WinRT APIs `PackageDependency.GetFromId()` and `PackageDependency.GetFromIdForSystem()` no longer fail for an unknown/invalid id e.g. `PackageDependency.GetFromId("not a valid id")`. Inappropriate later use will fail e.g.

```
var pd = PackageDependency.GetFromId("not a valid id");
var context = pd.Add();
// Add() fails and throws an exception
```

NOTE: A new OS API is needed to determine if a `PackageDependencyId` is a valid (known) value. WinAppSDK will be updated to use that when available.

https://task.ms/47326537

### CODE REVIEW

**Comment by kythant** [comment]
Just a heads up. Develop branch has been renamed to become the new Main Branch and is the default branch from now on.

**Comment by Scottj1s** [comment]
[celebrate]     Scott Jones reacted to your message:
________________________________

File: dev/DynamicDependency/API/M.AM.DD.PackageDependency.cpp [hunk 1]
```diff
@@ -25,21 +27,37 @@ namespace winrt::Microsoft::Windows::ApplicationModel::DynamicDependency::implem
 
     winrt::PackageDependency PackageDependency::GetFromId(hstring const& id)
     {
-        auto tokenUser{ wil::get_token_information<TOKEN_USER>(GetCurrentThreadEffectiveToken()) };
-        auto exists{ MddCore::PackageDependencyManager::ExistsPackageDependency(tokenUser->User.Sid, id.c_str()) };
-        if (!exists)
+        // Use the Win11 APIs if available (instead of Detour'ing to our own implementation)
+        if (MddCore::Win11::IsSupported())
+        {
+            //TODO:GetResolved2 GetPackageDependencyInfo2(userSid, id.c_str(), &packageDependencyInfo)
```
[the remark this answers is not in the document]
Comment (DrusTheAxe) [reply dev/DynamicDependency/API/M.AM.DD.PackageDependency.cpp:33]: There is now https://task.ms/47775758

Tweaked comments to reference it

File: dev/DynamicDependency/API/MsixDynamicDependency.h [hunk 2]
```diff
@@ -192,17 +192,30 @@ STDAPI MddAddPackageDependency(
...
     _In_ PCWSTR packageDependencyId,
     _Outptr_result_maybenull_ PWSTR* packageFullName) noexcept;
 
+/// Return the package full name to which the
+/// PackageDependency has been resolved. Does not add the
+/// package to the process graph.
+///
+/// @param packageFullName allocated via HeapAlloc; use HeapFree to deallocate.
+///                        If the package dependency cannot be resolved the function
+///                        succeeds but packageFullName is nullptr.
```
[the remark this answers is not in the document]
Comment (DrusTheAxe) [reply dev/DynamicDependency/API/MsixDynamicDependency.h:213]: Doh! Copy/paste error. Thanks

**Review by wcheng-msft — APPROVED** [review]
:shipit:

**Pull request merged**